crates/wast/src/parser.rs

```diff
@@ -630,6 +633,157 @@ impl<'a> Parser<'a> {
     pub fn cur_span(&self) -> Span {
         self.cursor().cur_span()
     }
+
+    /// Registers a new known annotation with this parser to allow parsing
+    /// annotations with this name.
+    ///
+    /// [WebAssembly annotations][annotation] are a proposal for the text format
+    /// which allows decorating the text format with custom structured
+    /// information. By default all annotations are ignored when parsing, but
+    /// the whole purpose of them is to sometimes parse them!
+    ///
+    /// To support parsing text annotations this method is used to allow
+    /// annotations and their tokens to *not* be skipped. Once an annotation is
+    /// registered with this method, then while the return value has not been
+    /// dropped (e.g. the scope of where this function is called) annotations
+    /// with the name `annotation` will be parse of the token stream and not
+    /// implicitly skipped.
+    ///
+    /// # Skipping annotations
+    ///
+    /// The behavior of skipping unknown/unregistered annotations can be
+    /// somewhat subtle and surprising, so if you're interested in parsing
+    /// annotations it's important to point out the importance of this method
+    /// and where to call it.
+    ///
+    /// Generally when parsing tokens you'll be bottoming out in various
+    /// `Cursor` methods. These are all documented as advancing the stream as
+    /// much as possible to the next token, skipping "irrelevant stuff" like
+    /// comments, whitespace, etc. The `Cursor` methods will also skip unknown
+    /// annotations. This means that if you parse *any* token, it will skip over
+    /// any number of annotations that are unknown at all times.
+    ///
+    /// To parse an annotation you must, before parsing any token of the
+    /// annotation, register the annotation via this method. This includes the
+    /// beginning `(` token, which is otherwise skipped if the annotation isn't
+    /// marked as registered. Typically parser parse the *contents* of an
+    /// s-expression, so this means that the outer parser of an s-expression
+    /// must register the custom annotation name, rather than the inner parser.
+    ///
+    /// # Return
+    ///
+    /// This function returns an RAII guard which, when dropped, will unregister
+    /// the `annotation` given. Parsing `annotation` is only supported while the
```

Comment:
Does anyone ever want to un-register an annotation?

Reply:
My thinking was that this is somewhat scope-based for a customizable parser like `wast` where while you're parsing a `module` you may want to strictly say what `@custom` means, but if you're parsing a `module` as part of a larger structure (e.g. `*.wast` files) you may want to assign a different meaning to `@custom` elsewhere.

Now this is all mega confusing if `@foo` means different things in different contexts so I doubt that anyone will do this in practice, but it felt better (to me at least) for these to be local decisions rather than global ones.

The only other thing I could think of was a sort of global registration API which stored global state, but that felt pretty bad and I figured it was best to avoid that.